lib/internal/async_hooks.js

```diff
@@ -369,31 +387,33 @@ function hasAsyncIdStack() {
 
 
 // This is the equivalent of the native push_async_ids() call.
-function pushAsyncIds(asyncId, triggerAsyncId) {
+function pushAsyncContext(asyncId, triggerAsyncId, resource) {
   const offset = async_hook_fields[kStackLength];
   if (offset * 2 >= async_wrap.async_ids_stack.length)
-    return pushAsyncIds_(asyncId, triggerAsyncId);
+    return pushAsyncContext_(asyncId, triggerAsyncId, resource);
   async_wrap.async_ids_stack[offset * 2] = async_id_fields[kExecutionAsyncId];
   async_wrap.async_ids_stack[offset * 2 + 1] = async_id_fields[kTriggerAsyncId];
+  execution_async_resources[offset] = resource;
   async_hook_fields[kStackLength]++;
   async_id_fields[kExecutionAsyncId] = asyncId;
   async_id_fields[kTriggerAsyncId] = triggerAsyncId;
 }
 
 
 // This is the equivalent of the native pop_async_ids() call.
-function popAsyncIds(asyncId) {
+function popAsyncContext(asyncId) {
   const stackLength = async_hook_fields[kStackLength];
   if (stackLength === 0) return false;
 
   if (enabledHooksExist() && async_id_fields[kExecutionAsyncId] !== asyncId) {
     // Do the same thing as the native code (i.e. crash hard).
-    return popAsyncIds_(asyncId);
+    return popAsyncContext_(asyncId);
   }
 
   const offset = stackLength - 1;
   async_id_fields[kExecutionAsyncId] = async_wrap.async_ids_stack[2 * offset];
   async_id_fields[kTriggerAsyncId] = async_wrap.async_ids_stack[2 * offset + 1];
+  delete execution_async_resources[offset];
```

Comment:
Could `execution_async_reesources` be a `Map` instead? Using `delete` is suboptimal for performance.

Reply:
@jasnell It’s an Array, not a regular object, so the perf implications are different. That being said, using an improvised benchmark, `.pop()` does seem to be faster than deleting the element here?

Reply:
Ah, I so rarely see `delete` used with arrays that I missed that 😆 ... `pop()` would definitely be better.

Reply:
I changed it to use `pop()` now. I opted for the delete before just for logical parity with what it does on the C++ side, but it's a somewhat performance-sensitive area so it makes sense to do whatever is fastest. :smile: